### DISCUSSION TITLE
how to Prevent delete data has relations?

### DISCUSSION THREAD

**Discussion by ha-akelius** [strapi/strapi#7793] [Legacy Topics]
what I did, I hooked the beforeDelete inside the model lifecycle, and it is working :) (user can't delete)
```
    async beforeDelete(params) {
      if (params._id) {
        const errors = await validateDeletePageId(params._id);
        if (!!errors) {
          throw new Error(errors);
        }
      }
    },
```
- but the user is seeing `An error occurred during record deletion` which is not good.
- now I want to improve this message, to show this entity has relations

**Comment by derrickmehaffy** [comment]
Is this within the Admin panel or some frontend application?

**Reply by ha-akelius** [reply]
yes I need it from Admin Panel, for 1 collection

**Reply by derrickmehaffy** [reply]
The way the lifecycles work, they are not good for returning error messages.

I would suggestion instead you create some kind of middleware, as there are 3 main places to handle this:
- In the backend, tweaking the error response sent to the frontend (requires maintaining some customization that could be difficult to migrate version to version
- In the frontend, again requires some customization that may be difficult to migrate
- A middleware that intercepts the traffic between the two and is easier to maintain

Here is an example, note this may not fit your exact use case but should point you in the direction of figuring out a good solution for you.

Path— `./config/middleware.js`

```js
module.exports = {
  load: {
    before: ['responseTime', 'logger', 'cors', 'responses', 'message'],
    order: [
      "Define the middlewares' load order by putting their name in this array in the right order",
    ],
    after: ['parser', 'router'],
  },
  settings: {
    message: {
      enabled: true
    }
  }
};
```

Path— `./middlewares/message/index.js`

```js
module.exports = (strapi) => {
  return {
    initialize() {
      strapi.app.use(async (ctx, next) => {
        await next();
        if (ctx.body.statusCode === 400 && ctx.request.url.startsWith('/content-manager/explorer')) {
          // analyse the ctx.body.data[0]
          // then transform the message to have an understandable one
          ctx.body.message[0] = {
            messages: [{
              id: 'My displayed message'
            }],
          };
        }
      });
    },
  };
};
```

**Reply by ha-akelius** [reply]
Thank I will try that 👍

**Reply by derrickmehaffy** [reply]
(Note this is only for the messages, it won't actually stop the request, and in fact when this is triggered is only after whatever action has happened)

**Reply by ha-akelius** [reply]
hi @derrickmehaffy 
sorry for the delay, but we have issues when upgrading from 3.0.x to 3.2.x

the solution is working 👍  after upgrading to version 3.2.x

thank you for your help

this is what I did: 

1. throwing error with errorMessage anywhere in my code, ex. `api/ANYMODEL/models/ANYMODEL.js`:
```

module.exports = {
  /**
   * Triggered before user creation.
   */
  lifecycles: {
    async beforeDelete(params) {
        const errors = await validate(params);
        if (!!errors) {
          const error = new Error(errors);
          error.errorMessage = errors;
          throw error;
        }
    },
   }
};
```

2. catch this error `errorMessage` in middleware and return badRequest:
```
module.exports = (strapi) => {
  return {
    initialize() {
      strapi.app.use(async (ctx, next) => {
        try {
          await next();
        } catch (error) {
          if (error.errorMessage) {
            ctx.badRequest(error.errorMessage);
          } else {
            throw error;
          }
        }
      });
    },
  };
};

```
